framework/src/meshgenerators/OverlayMeshGenerator.C

```diff
@@ -0,0 +1,140 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+#include "OverlayMeshGenerator.h"
+#include "CastUniquePointer.h"
+// libMesh includes
+#include "libmesh/mesh_generation.h"
+#include "libmesh/unstructured_mesh.h"
+#include "libmesh/replicated_mesh.h"
+#include "libmesh/mesh_modification.h"
+#include "libmesh/mesh_tools.h"
+
+registerMooseObject("MooseApp", OverlayMeshGenerator);
+InputParameters
+OverlayMeshGenerator::validParams()
+{
+  InputParameters params = MeshGenerator::validParams();
+  params.addClassDescription("This OverlayMeshGenerator creates a Cartesian mesh using "
+                             "DistributedRectilinearMeshGenerator in "
+                             "the mesh block region.");
+  params.addRequiredParam<MeshGeneratorName>("input", "The base mesh we want to overlay");
+  MooseEnum dims("1=1 2 3");
+  params.addRequiredParam<MooseEnum>(
+      "dim", dims, "The dimension of the mesh to be generated"); // Make this parameter required
+  params.addParam<dof_id_type>("nx", 1, "Number of elements in the X direction");
+  params.addParam<dof_id_type>("ny", 1, "Number of elements in the Y direction");
+  params.addParam<dof_id_type>("nz", 1, "Number of elements in the Z direction");
+  params.addParam<processor_id_type>(
+      "num_cores_for_partition",
+      0,
+      "Number of cores for partitioning the graph (dafaults to the number of MPI ranks)");
+  params.addRangeCheckedParam<unsigned>(
+      "num_side_layers",
+      2,
+      "num_side_layers>=1 & num_side_layers<5",
+      "Number of layers of off-processor side neighbors is reserved during mesh generation");
+  MooseEnum partition("graph linear square", "graph", false);
+  params.addParam<MooseEnum>(
+      "partition", partition, "Which method (graph linear square) use to partition mesh");
+  MooseEnum elem_types(
+      "EDGE EDGE2 EDGE3 EDGE4 QUAD QUAD4 QUAD8 QUAD9 TRI3 TRI6 HEX HEX8 HEX20 HEX27 TET4 TET10 "
+      "PRISM6 PRISM15 PRISM18 PYRAMID5 PYRAMID13 PYRAMID14"); // no default
```

Comment:
Should we be turning this into a `static const` declared in a header somewhere, rather than having a copy in `GeneratedMesh` and another in `ElementGenerator` and another in `GeneratedMeshGenerator` and another in `DistributedRectilinearMeshGenerator` and now another here?

Reply:
I follow the example of  `MooseEnum partitioning`  and add `elemTypes()` in `Moosemesh`. What do you think?

Reply:
Yeah, from your point of view I don't think the 4th sin is much worse than the 2nd or 3rd were.  Just wondering if anyone knows of a reason for the duplication or if we should get of it later.

Reply:
we should get rid of it. Sadly it is our burden as the maintainers of the pyramid to break the sinning cycles

Reply:
I am also curious if there is a better solution for this? `Moosemesh` has a partitioning function but the return value doesn't fit the need here. Similar situations happen in other mesh generators as well @loganharbour Any thoughts?